pkg/metricprovider/cron_metric_provider.go

```diff
@@ -0,0 +1,283 @@
+package metricprovider
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	autoscalingv2 "k8s.io/api/autoscaling/v2beta2"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog/v2"
+	"k8s.io/metrics/pkg/apis/external_metrics"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/custom-metrics-apiserver/pkg/provider"
+
+	autoscalingapi "github.com/gocrane/api/autoscaling/v1alpha1"
+	craneclientset "github.com/gocrane/api/pkg/generated/clientset/versioned"
+	craneinformers "github.com/gocrane/api/pkg/generated/informers/externalversions"
+)
+
+const (
+	// DefaultCronTargetMetricValue is used to construct a default external cron metric targetValue
+	DefaultCronTargetMetricValue int64 = 1
+	informerSyncPeriod                 = time.Minute
+)
+
+// CronMetricProvider implements ehpa cron metric as external metric provider
+type CronMetricProvider struct {
+	client               client.Client
+	craneInformerFactory craneinformers.SharedInformerFactory
+	ehpaInformer         cache.SharedIndexInformer
+
+	scalersLock sync.RWMutex
+	cronScalers map[string]*CronScaler
+}
+
+func NewCronMetricProvider(client client.Client, craneClient craneclientset.Interface) *CronMetricProvider {
+
+	craneInformerFactory := craneinformers.NewSharedInformerFactory(craneClient, informerSyncPeriod)
+	ehpaInformer := craneInformerFactory.Autoscaling().V1alpha1().EffectiveHorizontalPodAutoscalers().Informer()
+
+	c := &CronMetricProvider{
+		client:               client,
+		craneInformerFactory: craneInformerFactory,
+		ehpaInformer:         ehpaInformer,
+		cronScalers:          make(map[string]*CronScaler),
+	}
+	c.ehpaInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    c.addEHPA,
+		DeleteFunc: c.delEHPA,
+		UpdateFunc: c.updateEHPA,
+	})
+
+	return c
+}
+
+func (p *CronMetricProvider) Start(stopCh <-chan struct{}) bool {
+	go p.craneInformerFactory.Start(stopCh)
+	return cache.WaitForNamedCacheSync("cron-metric-provider", stopCh, p.ehpaInformer.HasSynced)
+}
+
+func (p *CronMetricProvider) addEHPA(obj interface{}) {
+	ehpa, ok := obj.(*autoscalingapi.EffectiveHorizontalPodAutoscaler)
```

Comment:
Your can referring to this controller pattern:
https://github.com/kubernetes/sample-controller/blob/master/controller.go

Reply:
> Your can referring to this controller pattern: https://github.com/kubernetes/sample-controller/blob/master/controller.go

ok